### ISSUE TITLE
Question: teaching one (or more) agent(s) to imitate another (online not offline)

### ISSUE THREAD

**Issue by pedropla** [ray-project/ray#4006]
I have a bit of a complex environment which sends many different types of (relatively large) data to the observable space. This requires a custom model with multiple inputs. The ideal model would likely contain a pretty deep network with convolutional and lstm layers. However, the time needed to train such a model with standard reinforcement learning on a machine with a couple of GPUs would likely run into weeks or months.

To test rllib for this environment I built a fairly simple MLP network with multiple inputs, etc and used curriculum learning to grow the environment size and complexity as the agent learned from the environment. After about 100-150M time steps it trains adequately for a small environment, but as the environment grows beyond a certain size, the model's performance declines significantly as it is understandably unable to generalize well.

What I would like to do is use this model as an agent to train another agent with a more complex model in the small/medium environment and once the other agent performs adequately in that environment then I would continue it's training using reinforcement learning.

I know offline datasets and training from that data is supported in rllib, but due to the observation space size and number of steps required, I would like to avoid that, as the files could get pretty huge and there would be a limitation as to how much of the environment can be replayed.

Is it possible with the current rllib implementation to have one agent learn from another with an online training run? For example, launch x workers with x environments each and then each environment would have 1 lead agent which would make the decisions and 1 or more agents learning from it in "real-time"? All should have access to the observation space and the lead agent's action space should be used to train the other agent(s).

The "ideal" would be to be able to train a few agents with different models at the same time from the same "lead agent" and then see which performs best and use that one. However, I would be happy to just be able to train one at a time for now.

Edit: Thinking it over a bit more, I think it might be best to have experiences collected from the lead agent with the multiple workers/environments per worker and then have a central process running in a loop training the new agent. Something like IMPALA, APE-X, APPO, etc, but with one agent generating experiences and another training from them. That would also allow training based on how well the lead agent performed on a given run.

**Comment by ericl** [comment]
Hey @pedropla , I think a good way of modeling this would be to use a multi-agent environment. I would imagine it working as follows:

- Let's assume you have a trained policy for the lead agent already.
- You wrap the real environment with a multi-agent env wrapper class. The wrapper env first sends each observation to the "lead agent" to get it's action. Then, it sends both the observation and the lead's action to the "student agent". The student agent looks at just the observation in its policy.
- The student agent has, in addition to its normal policy loss, an IL loss to imitate the teacher action. Over time perhaps this IL loss is annealed to zero.
- The lead agent does not learn (this can be done by omitting its policy id from "policies_to_train" in the multiagent config).

That way, the lead agent's actions are used to bootstrap the student policy. To get this working, you would have to write (1) the multi-agent wrapper for this teacher-student env, (2) a custom model for the student that only looks at the "real" component of its observation, and (3) a custom policy graph for the student with an IL loss based on the lead's action. Let me know if this approach makes sense.

A potentially simpler approach would be to allow the student to use both the real observation and teacher action in its policy (i.e. similar to progressive networks: https://arxiv.org/abs/1606.04671), but then you're stuck with the lead agent policy as part of the student unless you figure out some way of gradually removing it.

**Comment by pedropla** [comment]
Hi @ericl,

Thanks for the detailed response. It's definitely given me many ideas and a lot to think about. 

I was looking through the documentation about how to implement the multi-agent wrapper you suggested as that seemed the most appropriate solution for the problem I had. The progressive network solution you suggested also seemed interesting but I didn't want to drag the original model along so I was more inclined toward the multi-agent solution.

While looking into the multi-agent wrapper I wondered if a more generic and reusable solution could be implemented for the student/learner model that would not require passing the action space to the model but directly to the policy and would integrate the necessary policy change into existing policies somewhat automatically.

The way I would envision that working is something like this:

1) A "multi-agent teacher-student env" class that inherits from the multi-agent env wrapper class but allows specifying if an agent is a teacher or a student, if it is a student it would require referencing the teacher agent. That would allow multiple teachers with multiple followings of students. This class could either be called directly from the config or inherited further and modified as needed.

2) The Action space of the teacher agent would be passed directly to the policy of the student agent. I haven't created a custom policy yet so I'm not sure if there is something I'm missing here. 

I'm guessing you mentioned a custom model for this solution so that the action space of the teacher could be passed to the observation space of the student and then ignored in the model but passed to the policy. If that is the case I was thinking of passing it directly to the policy if that can be done.

3) Extending existing policies to add the additional IL loss you mention above. I have not modified or created a custom policy before so I'd have to see exactly how to do this. But if it could be standardised and reusable then it would be easier for others to implement in the future without having to create custom policies or models.

Of course if it is to be generalised it would have to have proper checks that the action spaces of both the teacher and students are the same, etc.

Do you think that a more generic solution such as this would be something useful to incorporate in rllib? Have I missed anything important due to my lack of in-depth knowledge of the rllib codebase?

Are there any similar examples of a custom policy that has an added loss but extends an existing policy without major changes that I could get some ideas from?

I see a few complications from the multi-agent solution though. The main one being that if we are using a standard loss policy in addition to the action space loss function on the student, that would require the environment to present a reward based on the students action and thus the student's environment would diverge from the teacher's environment. Unless the reward loss is only activated once the student agent is no longer a student.

The last solution you propose is possibly the easiest to implement but from my understanding of your explanation it seems more similar to the progressive network solution but with some manual gates to stop parts of the network at manually defined levels. 

Another "hack" I thought of based on your progressive network idea was integrating the teacher agent in the single agent environment and passing its action space directly to the observation space of the student agent and then using a loss function in the model (although I wasn't sure exactly how). But I feel the multi-agent approach with a teacher/student extension might be the cleanest and most reusable over a wider array of models/scenarios.

**Comment by ericl** [comment]
> If that is the case I was thinking of passing it directly to the policy if that can be done.

So you can definitely do that, in fact with a custom policy graph you can override the postprocess_trajectory() method to add whatever info from the teacher agent's batch you want, such as their actions. Then the child policy can use that in its loss definition. I guess that would avoid the need for a custom model.

There's an example of using postprocess_trajectory() for a centralized value function here, which is somewhat related: https://github.com/ray-project/ray/pull/3642/files#diff-96d0dc3c3c99e93a9161b96a810cbbbeR385

For customizing a policy graph, my general approach is to copy paste the entire class and modify it as needed, though in some cases you can inherit from the class only.

> Do you think that a more generic solution such as this would be something useful to incorporate in rllib? Have I missed anything important due to my lack of in-depth knowledge of the rllib codebase?

Maybe. If it can be built using existing APIs it would be even better. As you noted it depends on how often the use case comes up as to whether it makes sense to have it baked in vs available as an example.

**Comment by pedropla** [comment]
Hi @ericl,

The last comment you made on your initial reply (about using a single model with manually gated components and a self-supervised loss) got me thinking. What if I could forgo modifying the policy graph at all and instead do everything through the standard reward system?

I wrote a really rough draft prototype to test if that idea was even viable. In the code below I first train the "teacher" agent and then set up the multi-agent environment and train the student based on whether its actions match the teacher. 

There is of course a major problem with using rewards for this. By design, the policy will calculate the rewards of not just the step which returns the rewards. This will reduce training efficiency as the training progresses. Is there any config flag that can be set so that the training policy only looks at the rewards for the step which returned them when calculating the loss function? If so, this might be a viable solution.

I am trying to make this as simple as possible and avoiding having to customize a policy so that it can be used across the different policies rllib supports and with whatever models (custom or not).

```
import gym

import ray
from ray.rllib.agents.ppo import PPOAgent
from ray.rllib.agents.ppo.ppo_policy_graph import PPOPolicyGraph
from ray.tune.logger import pretty_print


from ray.rllib.env.multi_agent_env import MultiAgentEnv

class TeacherStudentEnv(MultiAgentEnv):
    def __init__(self, config_env={}, env="CartPole-v0"):
        self.env = gym.make(env)
        self.observation_space = self.env.observation_space
        self.action_space = self.env.action_space
        
    def reset(self):
        obs = self.env.reset()
        return {"teacher": obs, "student": obs}

    def step(self, action_dict):
        obs, rew, done, info = self.env.step(action_dict["teacher"])
        student_reward = self._get_student_reward(action_dict["teacher"], action_dict["student"])
        return {"teacher":obs, "student":obs}, \
            {"teacher":rew, "student":student_reward}, \
            {"teacher":done, "student":done, "__all__":done}, \
            {"teacher":info, "student":info}
    
    # would only work with discrete actions, would need extending for continuous action spaces
    def _get_student_reward(self, teacher_action, student_action):
        return 1 if student_action == teacher_action else -1
        
        
        
# train the teacher model
def train_teacher(env, target_reward_mean=200):
    agent = PPOAgent(env=env)
    result={'episode_reward_mean': 0}
    while result["episode_reward_mean"] < target_reward_mean:
        result = agent.train()
        print(pretty_print(result))
    weights=agent.get_policy().get_weights()
    return weights

if __name__ == "__main__":
    ray.init()
    env_name="CartPole-v0"
    teacher_weights=train_teacher(env_name)
    
    multi_env=TeacherStudentEnv(env_name)
    
    def train(config, reporter):
        agent = PPOAgent(config=config)
        global teacher_weights
        agent.local_evaluator.policy_map["teacher_policy"].set_weights(teacher_weights)
        result=None
        while True:
            result=agent.train()
            print(pretty_print(result))
        return result
        

    def policy_mapping_fn(agent_id):
        if agent_id.startswith("teacher"):
            return "teacher_policy"
        else:
            return "student_policy"

    ray.tune.run_experiments({
        "teacher_student": {
            "run": train, 
            "env": TeacherStudentEnv,
            "config": {
                "num_workers": 0,
                "log_level": "INFO",
                "multiagent": {
                    "policy_graphs": {
                        "teacher_policy": (PPOPolicyGraph,
                                              multi_env.observation_space,
                                              multi_env.action_space, {}),
                        "student_policy": (PPOPolicyGraph,
                                              multi_env.observation_space,
                                              multi_env.action_space, {}),
                    },
                    "policy_mapping_fn": ray.tune.function(policy_mapping_fn),
                    "policies_to_train": ["student_policy"]
                },
            },
        },
    })

```

**Comment by ericl** [comment]
That's a neat idea. I think you can set gamma=0 to handle this?

Another challenge is that you'd want to optimize then cross entropy of the
action distribution probably, and not just the exact action match.

**Comment by pedropla** [comment]
Gamma=0 is what I needed.

Cross entropy will be a bit more challenging as with discrete actions the environment doesn't have (to my knowledge) access to the probability distribution that came from the model. For continuous actions it might be easier to come up with a reward that is scaled based on how far off the values are. 

For my particular use case I have a large observation space and a binary action space so the simple solution of an exact match reward might suffice.

**Comment by edoakes** [comment]
Stale - please open new issue if still relevant
